### PR TITLE
Add `AvgPool` derivative tests

### PR DESCRIPTION
**Pull request by jon-tow**
#402

### CODE REVIEW

File: Tests/TensorFlowTests/LayerTests.swift [hunk 1]
```diff
@@ -368,6 +388,26 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testAvgPool2DGradient(){
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]:
```suggestion
    func testAvgPool2DGradient() {
```

File: Tests/TensorFlowTests/LayerTests.swift [hunk 2]
```diff
@@ -360,6 +360,26 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testAvgPool1DGradient(){
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]:
```suggestion
    func testAvgPool1DGradient() {
```

File: Tests/TensorFlowTests/LayerTests.swift [hunk 3]
```diff
@@ -376,6 +416,22 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testAvgPool3DGradient(){
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]:
```suggestion
    func testAvgPool3DGradient() {
```

File: Tests/TensorFlowTests/LayerTests.swift [hunk 4]
```diff
@@ -360,6 +360,26 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testAvgPool1DGradient() {
+        let layer = AvgPool1D<Float>(poolSize: 2, stride: 1, padding: .valid)
+        let x = Tensor(shape: [1, 4, 4], scalars: (0..<16).map(Float.init))
+        let computedGradient = gradient(at: x, layer) { $1($0).sum() }
+        // The expected value of the gradient was computed using the following Python code:
+        // ```
+        //   avgpool1D = tf.keras.layers.AvgPool1D(strides=1)
```
Comment (Shashi456) [inline Tests/TensorFlowTests/LayerTests.swift]: You haven't mentioned the poolsize for the layer, I'm not sure if this will work. Same for the `avgpool2d` and `avgpool3d` example.

Comment (jon-tow) [reply Tests/TensorFlowTests/LayerTests.swift]: @Shashi456  The default arguments of `pool_size` are `2`, `(2, 2)`, and `(2, 2, 2)` for the respective 1D, 2D, and 3D layers. At least this is the case with keras under tf 2.0.

**Comment by jon-tow** [comment]
I thought I pulled upstream recently enough. This should work correctly now.

**Comment by Shashi456** [comment]
@jon-tow if you merge into the master, this will pass.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 5]
```diff
@@ -360,6 +360,26 @@ final class LayerTests: XCTestCase {
...
+        //   with tf.GradientTape() as t:
+        //     t.watch(x)
+        //     y = tf.math.reduce_sum(avgpool1D(x))
+        //   print(t.gradient(y, x))
+        // ```
+        let expectedGradient = Tensor<Float>([[
+             [0.5, 0.5, 0.5, 0.5],
```
Comment (Shashi456) [inline Tests/TensorFlowTests/LayerTests.swift]: 4 space indentation here.

**Comment by jon-tow** [comment]
Thanks @Shashi456. Just finalized everything now. Appreciate the helpful eye :+1:
